File: 17 Letter Combinations of a Phone Number.py
class Solution:
    def __init__(self):
        self.out = []
        self.ins = ""
        self.table = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
    def _get(self, s, index):
        if index == len(self.ins):
            self.out.append(s);
        else:
            for i in range(len(self.table[ord(self.ins[index]) - ord('0')])):
                self._get(s + self.table[ord(self.ins[index]) - ord('0')][i], index + 1)
    def _getstr(self):
        self._get("", 0);

    def letterCombinations(self, digits):
        if digits == "":
            return self.out
        self.ins = digits
        self._getstr()
        return self.out

File: 17 Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        out.resize(0);
        if(!digits.size())
            return out;
        in = digits;
        getstr();
        return out;
    }
private:
    void getstr(){
        get("", 0);
    }
    void get(string s, int index){
        if(index == in.size()){
            out.push_back(s);
        }
        else{
            for(int i = 0;i < table[in[index] - '0'].size();++i){
                get(s + table[in[index] - '0'][i], index + 1);
            }
        }
    }
    vector<string> out;
    string in;
    string table[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
};